File: 03_code.cpp
#include <algorithm>
#include <iostream>
#include <fstream>
#include <string>
#include <regex>
#include <sstream>
#include <vector>

int main() {
    const std::string filename = "03_input.txt";

    std::ifstream file(filename, std::ios::in);
    if (!file.is_open()) {
        std::cerr << "Failed to open the file." << std::endl;
        return 1;
    }

    std::string content;
    std::string line;
    while (std::getline(file, line)) {
        content += line;
    }

    std::regex pattern(R"((don't\(\))|(do\(\))|mul\((\d{1,3}),(\d{1,3})\))");
    std::vector<std::tuple<bool, bool, std::string, std::string>> results;

    int total_sum_1 = 0;
    int total_sum_2 = 0;

    auto words_begin = std::sregex_iterator(content.begin(), content.end(), pattern);
    auto words_end = std::sregex_iterator();

    for (std::regex_iterator i = words_begin; i != words_end; ++i) {
        std::smatch match = *i;
        bool disable = !match[1].str().empty();
        bool enable = !match[2].str().empty();
        std::string a = match[3].str();
        std::string b = match[4].str();
        results.emplace_back(disable, enable, a, b);
    }

    bool is_enabled = true;

    for (const auto& [disable_flag, enable_flag, a, b] : results) {
        if (!a.empty() && !b.empty()) {
            int product = std::stoi(a) * std::stoi(b);
            total_sum_1 += product;
            if (is_enabled) total_sum_2 += product;
        }
        if (disable_flag) is_enabled = false;
        if (enable_flag) is_enabled = true;
    }

    file.close();

    std::cout << "First part:" << total_sum_1 << std::endl;
    std::cout << "Second part:" << total_sum_2 << std::endl;

    return 0;
}

File: 03_code.py
import re
input_file = "03_input.txt"
test_input = "xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))"
long_string = ""
with open(input_file, "r") as f:
    for line in f.readlines():
        long_string += line

pattern = re.compile(r"(don't\(\))|(do\(\))|mul\((\d{1,3}),(\d{1,3})\)" )
total_sum_1 = 0
total_sum_2 = 0
result = pattern.findall(long_string)
enabled = True
for matches in result:
    disable_flag = matches[0]
    enable_flag = matches[1]
    a = matches[2]
    b = matches[3]

    if len(a) and len(b):
        assert len(disable_flag) == 0
        assert len(enable_flag) == 0
        prod = int(a) * int(b)
        total_sum_1 += prod
        if enabled:
            total_sum_2 += prod

    if len(disable_flag):
        assert len(a) == 0
        assert len(b) == 0
        assert len(enable_flag) == 0
        enabled = False

    if len(enable_flag):
        assert len(a) == 0
        assert len(b) == 0
        assert len(disable_flag) == 0
        enabled = True


print(f"First part: {total_sum_1}")
print(f"Second part: {total_sum_2}")
